rust_snuba/rust_arroyo/src/backends/kafka/config.rs

```diff
@@ -11,129 +11,129 @@ pub struct OffsetResetConfig {
 
 #[derive(Debug, Clone)]
 pub struct KafkaConfig {
-    config_map: HashMap<String, String>,
-    // Only applies to consumers
-    offset_reset_config: Option<OffsetResetConfig>,
+    bootstrap_servers: Vec<String>,
+    extra_config: HashMap<String, String>,
 }
 
 impl KafkaConfig {
-    pub fn new_config(
-        bootstrap_servers: Vec<String>,
-        override_params: Option<HashMap<String, String>>,
-    ) -> Self {
-        let mut config_map = HashMap::new();
-        config_map.insert("bootstrap.servers".to_string(), bootstrap_servers.join(","));
-        let config = Self {
-            config_map,
-            offset_reset_config: None,
-        };
+    pub fn new(bootstrap_servers: Vec<String>, extra_config: HashMap<String, String>) -> Self {
+        Self {
+            bootstrap_servers,
+            extra_config,
+        }
+    }
+}
 
-        apply_override_params(config, override_params)
+impl From<KafkaConfig> for RdKafkaConfig {
+    fn from(cfg: KafkaConfig) -> Self {
+        let mut config_obj = RdKafkaConfig::new();
+
+        config_obj.set(
+            "bootstrap.servers".to_string(),
+            cfg.bootstrap_servers.join(","),
+        );
+
+        for (key, val) in cfg.extra_config.into_iter() {
+            config_obj.set(key, val);
+        }
+
+        config_obj
     }
+}
 
-    pub fn new_consumer_config(
+#[derive(Debug, Clone)]
+pub struct KafkaConsumerConfig {
+    pub core: KafkaConfig,
+    pub group_id: String,
+    pub max_poll_interval_ms: usize,
+    pub session_timeout_ms: Option<usize>,
```

Comment:
This comment made me realize that this design is kinda unsound: what happens if you construct a `KafkaConfig` with e.g. a `session_timeout_ms` value in the `extra_config`, but then create a `KafkaConsumerConfig` from that `KafkaConfig` that sets its `session_timeout_ms` field? You'd have two inconsistent values. The same thing can happen with any field on `KafkaConsumerConfig` because of the open-endedness of `extra_config`.

Reply:
You're right, it's probably not worth the hassle.